crates/forge/bin/cmd/clone.rs

```diff
@@ -0,0 +1,728 @@
+    for entry in std::fs::read_dir(tmp_dump_dir.join(contract_name))? {
+        if std::fs::metadata(&src_dir).is_err() {
+            std::fs::create_dir(&src_dir)?;
+        }
+        let entry = entry?;
+        let folder_name = entry.file_name().to_string_lossy().to_string();
+        // special handling for contracts and src directories: we flatten them.
+        if strip_old_src && (folder_name.as_str() == "contracts" || folder_name.as_str() == "src") {
+            // move all sub folders in contracts to src
+            for e in read_dir(entry.path())? {
+                let e = e?;
+                let dest = src_dir.join(e.file_name());
+                std::fs::rename(e.path(), &dest)?;
+                remappings.push(Remapping {
+                    context: None,
+                    name: format!("{}/{}", folder_name, e.file_name().to_string_lossy()),
+                    path: dest.to_string_lossy().to_string(),
+                });
+            }
+        } else {
+            // move the other folders to src
+            let dest = src_dir.join(entry.file_name());
+            std::fs::rename(entry.path(), &dest)?;
+            remappings.push(Remapping {
+                context: None,
+                name: entry.file_name().to_string_lossy().to_string(),
+                path: dest.to_string_lossy().to_string(),
+            });
+        }
+    }
+
+    // remove the temporary directory
+    std::fs::remove_dir_all(tmp_dump_dir)?;
+
+    Ok((remappings.into_iter().map(|r| r.into_relative(root)).collect(), strip_old_src))
+}
+
+/// Compile the project in the root directory, and return the compilation result.
+/// XXX (ZZ): disable output when the quite flag is set.
+pub fn compile_project(root: &PathBuf, quiet: bool) -> Result<ProjectCompileOutput> {
+    let mut config = Config::load_with_root(root);
+    config.extra_output.push(ContractOutputSelection::StorageLayout);
+    let project = config.project()?;
+    let compiler = ProjectCompiler::new().quiet_if(quiet);
+    compiler.compile(&project)
+}
+
+/// Find the artifact of the contract with the specified name.
+/// This function returns the path to the source file and the artifact.
+pub fn find_main_contract<'a>(
+    compile_output: &'a ProjectCompileOutput,
+    contract: &str,
+) -> Result<(PathBuf, &'a ConfigurableContractArtifact)> {
+    // XXX (ZZ): it is possible that we have multiple contracts with the same name
+    // in different files, we should handle this case in the future
+    for (f, c, a) in compile_output.artifacts_with_files() {
+        if contract == c {
+            return Ok((PathBuf::from(f), a));
+        }
+    }
+    Err(eyre::eyre!("contract not found"))
+}
+
+#[cfg(test)]
+use mockall::automock;
+/// EtherscanClient is a trait that defines the methods to interact with Etherscan.
+/// It is defined as a wrapper of the `foundry_block_explorers::Client` to allow mocking.
+#[cfg_attr(test, automock)]
+pub(crate) trait EtherscanClient {
```

Comment:
can we move this into the tests module?

Reply:
This trait is used by the function `run_with_client`, so we cannot move this trait into the test module because otherwise there will be `trait not found` compilation errors in the cargo build. The rationale of this trait is to allow mocking for Etherscan clients. I have made my best to use `#[cfg(test)]` and `#[cft_attr(test, automock)]` so that the mocking is only compiled in `test` build target.

This is the recommended practice in the doc of `mockall`: https://github.com/asomers/mockall?tab=readme-ov-file#usage